### PR TITLE
feat: add drag-and-drop reordering to the word list

### PR DESCRIPTION
**Pull request by Vladyslav-Soldatenko**
## Description

Adds long-press-free, drag-handle-based reordering to the word list. Dragging writes the same `position` column that already drives both list display order and (under the default "Sequential" new-card setting) which word is introduced next in study, so manually reordering the list is a deliberate way to curate study order too.

This closes a related pre-existing correctness gap: deleting a word previously left a permanent gap in `position` values. Delete (single and bulk) now renumbers the remaining rows contiguously in the same transaction, and a one-time migration (`MIGRATION_5_6`) cleans up any gaps already present in existing installs.

## Type of Change

- [x] New feature

## Changes Made

- Domain: `VocabularyItem` gains a `position` field; `VocabularyCatalogRepository.reorderVocabulary(orderedIds)` added.
- Data: `VocabularyDao.reorderVocabulary`/`deleteVocabularyAndRenumber` (two-phase negative/positive position writes to avoid transient UNIQUE-index collisions); `MIGRATION_5_6` for legacy gap cleanup (DB version 5 → 6).
- ViewModel: `WordListViewModel.reorderWords` guards against sub-2-item lists, no-op reorders, and stale/foreign id sets before writing; swallows a failed write (logged nowhere else in this VM either) rather than crashing, since it's a low-stakes convenience action.
- UI: a leading-edge drag handle per row (`sh.calvin.reorderable` 3.1.0), hidden during search or multi-select. Long-press-to-select is completely unchanged.
- New dependency: `sh.calvin.reorderable:reorderable:3.1.0` (Apache 2.0).

## How to Test

1. Open the word list with 2+ words, drag a row by its leading handle to a new position, confirm it persists after navigating away and back.
2. Confirm the handle disappears while searching or in multi-select mode, and that long-press still enters multi-select as before.
3. Delete a word, confirm no gap is left in the remaining words' effective order (visible via re-adding a word - it should slot in immediately after the last one, not skip a number).

## Checklist

- [x] Code follows the project's style guidelines
- [x] Self-reviewed the code for obvious errors
- [x] Added or updated tests where applicable, including edge cases
- [x] Existing tests pass locally
- [ ] Updated documentation if needed
- [x] No duplication introduced. I searched the existing code for similar behavior and extracted similar parts to be reused
- [ ] No new warnings or supressions introduced. If something like this has to be introduced - get approve from maintainer first

**Flagging the suppressions explicitly per the checklist above** (all detekt, none disable a correctness check):
- `WordListViewModel.reorderWords`: `TooGenericExceptionCaught`/`SwallowedException` - deliberate broad catch so a failed reorder write never crashes the app; documented inline.
- `VocabularyRepositoryImpl`: `TooManyFunctions` - implements two domain interfaces (`VocabularyCatalogRepository` + `VocabularyStudyRepository`) that share DAOs/transactions.
- `AppDatabaseMigrationTest`: `LargeClass` - one independent test per historical migration/edge case, matching the file's existing structure.
- `WordListContent`: added `CyclomaticComplexMethod` to its existing screen-level-state suppression (already suppresses `LongMethod`/`CognitiveComplexMethod` for the same reason).

## Test coverage

- New DAO unit tests (`VocabularyDaoReorderTest`) covering reorder/delete-renumber edge cases: empty/single-item input, no-op, first↔last, duplicate/unknown/partial id sets (including the one that's expected to throw and roll back), non-position fields preserved, gap-closing on delete.
- New `MIGRATION_5_6` tests in `AppDatabaseMigrationTest` (gapped legacy data, already-contiguous data, empty table, relative order, non-position columns, unique-index still enforced).
- `VocabularyMapperTest`, `WordListViewModelTest` additions for the new field/method.
- `WordListScreenTest` (Robolectric) additions for handle visibility gating and gesture → callback behavior.
- New instrumented e2e suite `WordListReorderE2eTest`: drag persistence across activity recreation, availability gating (search/selection/single-word), rotation-mid-drag safety, auto-scroll on a long list, interaction with sequential new-card scheduling, delete, and Anki import.

I can't run instrumented tests in this environment, so I'll be watching this PR's CI run and pushing fixes if the e2e suite surfaces anything.

---
_Generated by [Claude Code](https://claude.ai/code/session_01GkjecuN3GxP3CQfEPtt365)_

### CODE REVIEW

File: app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt [hunk 1]
```diff
@@ -73,6 +76,30 @@ interface VocabularyDao {
...
+        id: Long,
+        position: Long,
+    )
+
+    // Two-phase write: negative placeholders first, then real 1-based positions. No two rows
+    // are ever assigned the same value at the same instant, so this can never trip the UNIQUE
+    // index on position regardless of the rows' current on-disk values. Callers must pass the
+    // complete current id set - a partial list leaves excluded rows' old positions untouched,
+    // which can collide with a newly-assigned value and abort the whole transaction.
+    @Transaction
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt:85]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/local/database/Migrations.kt [hunk 2]
```diff
@@ -93,3 +93,28 @@ val MIGRATION_4_5 =
...
     }
+
+// Pure data-fix migration: closes any gaps in `position` left by deletions before this
+// version (deletion now renumbers remaining rows on its own; this is a one-time catch-up for
+// data that predates that fix). No column/index change, so schema-wise this is a no-op -
+// still requires a version bump for Room's migration system to run it. Same two-phase
+// negative-then-positive renumber the DAO uses, done procedurally (not a SQL window function)
+// to avoid depending on any particular on-device SQLite version's feature set for a migration
+// that only ever runs once, on a small table.
+val MIGRATION_5_6 =
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/local/database/Migrations.kt:97]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/repository/VocabularyRepositoryImpl.kt [hunk 3]
```diff
@@ -52,6 +52,11 @@ internal fun todayStamp(): Int =
 class NoAvailableItemsException : Exception("Daily limits reached and no reviews due")
 
 @Singleton
+// Function count comes from implementing two domain interfaces
+// (VocabularyCatalogRepository + VocabularyStudyRepository) in one class, not from an
+// undecomposed monolith - splitting further would separate methods that share the same DAOs
+// and transactional invariants.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/repository/VocabularyRepositoryImpl.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/ui/WordListViewModel.kt [hunk 4]
```diff
@@ -98,6 +99,30 @@ class WordListViewModel
             }
         }
 
+        // The set-equality guard is a correctness guard, not just tidiness: it's what keeps
+        // repository.reorderVocabulary from ever being called with a partial id set, which
+        // would leave excluded rows' stale positions colliding with the newly-assigned ones.
+        // Reordering is a low-stakes convenience action, so unlike every other method here, a
+        // repository failure is swallowed rather than left to propagate - the UI already
+        // reflects the attempted order locally and will resync to the last-persisted order on
+        // the next `words` emission.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/ui/WordListViewModel.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/ui/WordListViewModel.kt [hunk 5]
```diff
@@ -98,6 +99,30 @@ class WordListViewModel
...
+            if (orderedIds.size < 2 || currentIds.size < 2) return
+            if (orderedIds.toSet() != currentIds.toSet()) return
+            if (orderedIds == currentIds) return
+            viewModelScope.launch {
+                try {
+                    repository.reorderVocabulary(orderedIds)
+                } catch (e: CancellationException) {
+                    throw e
+                } catch (e: Exception) {
+                    // Swallowed intentionally - see comment above.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/ui/WordListViewModel.kt]: remove comment

File: app/src/test/java/com/procrastilearn/app/data/local/dao/VocabularyDaoReorderTest.kt [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+                incorrectCount = incorrectCount,
+                position = position,
+            ),
+        )
+
+    private suspend fun entityById(id: Long): VocabularyEntity = requireNotNull(dao.getVocabularyById(id))
+
+    private suspend fun positionOf(id: Long): Long = entityById(id).position
+
+    // --- reorderVocabulary ---
```
Comment (Vladyslav-Soldatenko) [inline app/src/test/java/com/procrastilearn/app/data/local/dao/VocabularyDaoReorderTest.kt]: remove comment

File: app/src/test/java/com/procrastilearn/app/data/local/dao/VocabularyDaoReorderTest.kt [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+        runTest {
+            val ids = (1..5).map { insert("word-$it") }
+
+            dao.reorderVocabulary(listOf(ids[3], ids[0], ids[4], ids[1], ids[2]))
+
+            val positions = ids.map { positionOf(it) }
+            assertThat(positions.distinct()).hasSize(positions.size)
+        }
+
+    // --- deleteVocabularyAndRenumber ---
```
Comment (Vladyslav-Soldatenko) [inline app/src/test/java/com/procrastilearn/app/data/local/dao/VocabularyDaoReorderTest.kt]: remove comment

File: app/src/test/java/com/procrastilearn/app/ui/screens/WordListScreenTest.kt [hunk 8]
```diff
@@ -649,4 +656,115 @@ class WordListScreenTest {
...
+    // and its not-yet-reached target don't move until the drag actually reaches them).
+    private fun centerYPx(node: SemanticsNodeInteraction): Float {
+        val bounds = node.fetchSemanticsNode().boundsInRoot
+        return (bounds.top + bounds.bottom) / 2f
+    }
+
+    // Landing exactly on the target row's original center is one swap short in practice (the
+    // dragged item needs to cross past a row's threshold, not just reach its midpoint), so aim
+    // well beyond it rather than exactly at it.
+    private fun stepDeltaYTowardItem(
```
Comment (Vladyslav-Soldatenko) [inline app/src/test/java/com/procrastilearn/app/ui/screens/WordListScreenTest.kt]: remove comment

File: app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+        repeat(DRAG_STEP_COUNT / 2) {
+            handle.performTouchInput { moveBy(Offset(0f, stepDeltaY)) }
+            composeTestRule.waitForIdle()
+        }
+        // A configuration change delivers ACTION_CANCEL to any in-flight touch sequence before
+        // tearing the view hierarchy down - simulate that rather than leaving the gesture
+        // truly dangling across the recreate() call. waitForIdle() gives the cancel a chance to
+        // actually propagate (and be observed as a no-commit) before recreate() tears everything
+        // down.
+        handle.performTouchInput { cancel() }
```
Comment (Vladyslav-Soldatenko) [inline app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt]: remove comment

File: app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+        navigateToWordList()
+        composeTestRule.waitUntilNodeExists(hasTestTag(itemTag(firstId)), DEFAULT_TIMEOUT_MS)
+
+        // A fixed absolute offset is a guess at where the library's auto-scroll trigger zone
+        // actually is, and can easily overshoot past the LazyColumn's own bounds entirely (into
+        // system UI below it), which never counts as "still inside the scrollable area." Aim at
+        // the bottom edge of whatever is currently visible instead - inset slightly so the point
+        // is guaranteed to land inside the last visible row, i.e. genuinely inside the list and
+        // as close to its trailing edge as content allows. Re-measure every step rather than
+        // once, since that edge itself moves as the list scrolls.
```
Comment (Vladyslav-Soldatenko) [inline app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt]: remove comment

File: app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+        val approachStepDeltaY = (initialEdgeTargetY - startY) / DRAG_STEP_COUNT
+
+        handle.performTouchInput { down(center) }
+
+        // A single moveTo() straight from the start position to the (far-away) edge target
+        // never registers as a drag at all - as with every other drag gesture in this suite,
+        // the gesture detector needs a series of small, incremental moves to recognize it as a
+        // continuous drag rather than a discontinuous jump.
+        repeat(DRAG_STEP_COUNT) {
+            handle.performTouchInput { moveBy(Offset(0f, approachStepDeltaY)) }
```
Comment (Vladyslav-Soldatenko) [inline app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt]: remove comment

File: app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+            composeTestRule.waitForIdle()
+        }
+
+        // Once near the edge, hold there so the auto-scroll loop has real wall-clock time to
+        // advance. Repeating moveTo() at the exact same coordinate produces a zero-delta pointer
+        // event, which the drag-tracking code has no obligation to treat as "still held near the
+        // edge" - alternate a tiny jitter each step so every event is a genuine, distinct motion.
```
Comment (Vladyslav-Soldatenko) [inline app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt]: remove comment

File: app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+
+        // For a 40-item list, only currently-visible rows are composed into the semantics tree,
+        // and where the list happens to be scrolled to isn't something this test controls - so
+        // read the persisted position directly from the DB rather than trying to re-locate a
+        // specific row (possibly off-screen) after the drop. Exact final position depends on
+        // real auto-scroll speed/timing, so keep the assertion loose: it only needs to have
+        // moved well past a simple adjacent swap, which is only reachable if auto-scroll
+        // actually engaged during the held drag (with no auto-scroll the drag could never reach
+        // past whatever was on-screen at the start).
+        val finalPosition = positionOf(firstId)
```
Comment (Vladyslav-Soldatenko) [inline app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt]: remove comment

File: app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+        dragHandleToItem(importedIds[1], existingId)
+
+        recreateActivity()
+        composeTestRule.waitUntilNodeExists(hasTestTag(itemTag(existingId)), DEFAULT_TIMEOUT_MS)
+        assertEquals(listOf(importedIds[1], existingId, importedIds[0]), displayedWordIdsInOrder())
+    }
+
+    // Activity.recreate() must run on the main thread - calling it directly from the
+    // instrumentation thread throws IllegalStateException.
+    private fun recreateActivity() {
```
Comment (Vladyslav-Soldatenko) [inline app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt:250]: remove comment

File: app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+        )
+        composeTestRule.onNodeWithText(string(R.string.action_delete)).performClick()
+        composeTestRule.waitForIdle()
+    }
+
+    // A single large moveTo() jump doesn't reliably register as a real drag - the target
+    // position can land outside the list's actual content bounds entirely, which the library
+    // treats as "no valid drop target" rather than clamping to the nearest one. Real on-screen
+    // row positions (not a guessed pixel distance) split into several incremental moveBy()
+    // steps, with idle time between each for layout to catch up, is what actually works.
```
Comment (Vladyslav-Soldatenko) [inline app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt]: remove comment

File: app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+        targetItemWordId: Long,
+    ): Float {
+        val handleCenterY = centerYPx(handleNode)
+        val targetCenterY = centerYPx(composeTestRule.onNodeWithTag(itemTag(targetItemWordId)))
+        return (targetCenterY - handleCenterY) / DRAG_STEP_COUNT
+    }
+
+    // Landing exactly on the target row's original center is one swap short in practice, so
+    // aim well beyond it rather than exactly at it.
+    private fun dragHandleToItem(
```
Comment (Vladyslav-Soldatenko) [inline app/src/androidTest/java/com/procrastilearn/app/e2e/WordListReorderE2eTest.kt]: remove comment
